include/mbedtls/bignum.h

```diff
@@ -186,512 +186,618 @@ typedef struct mbedtls_mpi
- *                 '1' bit
+ * \brief          Return the number of zero-bits before the
+ *                 least significant 1-bit
  *
- * Note: Thus also the zero-based index of the least significant '1' bit
+ * \note           This is the same as the zero-based index of
+ *                 the least significant '1' bit.
  *
- * \param X        MPI to use
+ * \param X        The MPI to query.
+ *
+ * \return         The number of zero-bits before the least significant
+ *                 1-bit in \p X.
  */
 size_t mbedtls_mpi_lsb( const mbedtls_mpi *X );
 
 /**
  * \brief          Return the number of bits up to and including the most
- *                 significant '1' bit'
+ *                 significant 1-bit.
+ *
+ * * \note         This is same as the one-based index of the most
+ *                 significant 1-bit.
  *
- * Note: Thus also the one-based index of the most significant '1' bit
+ * \param X        The MPI to query. Must point to an initialized MPI.
  *
- * \param X        MPI to use
+ * \return         The number of bits up to and including the most
+ *                 significant 1-bit.
  */
 size_t mbedtls_mpi_bitlen( const mbedtls_mpi *X );
 
 /**
- * \brief          Return the total size in bytes
+ * \brief          Return the total size an MPI value in bytes.
  *
- * \param X        MPI to use
+ * \param X        The MPI to use. Must point to an initialized MPI.
+ *
+ * \note           The value returned by this function may be less than
+ *                 the number of bytes used to store \p X internally.
+ *                 This happens if and only if there are trailing zero-bytes.
+ *
+ * \return         The least number of bytes capable of storing
+ *                 the absolute value of \p X.
  */
 size_t mbedtls_mpi_size( const mbedtls_mpi *X );
 
 /**
- * \brief          Import from an ASCII string
+ * \brief          Import an MPI from an ASCII string.
  *
- * \param X        Destination MPI
- * \param radix    Input numeric base
- * \param s        Null-terminated string buffer
+ * \param X        The destination MPI. Must point to an initialized MPI.
+ * \param radix    The numeric base of the input string.
+ * \param s        Null-terminated string buffer.
  *
- * \return         0 if successful, or a MBEDTLS_ERR_MPI_XXX error code
+ * \return         \c 0 if successful.
+ * \return         A negative error code on failure.
  */
 int mbedtls_mpi_read_string( mbedtls_mpi *X, int radix, const char *s );
 
 /**
- * \brief          Export into an ASCII string
+ * \brief          Export an MPI into an ASCII string.
  *
- * \param X        Source MPI
- * \param radix    Output numeric base
- * \param buf      Buffer to write the string to
- * \param buflen   Length of buf
- * \param olen     Length of the string written, including final NUL byte
+ * \param X        The source MPI. Must point to an initialized MPI.
+ * \param radix    The numeric base of the output string.
```

Comment:
I intentionally omitted this here, because I don't think that should be a precondition. A precondition is something we make absolutely no guarantees for if it's violated; different from an implementation limitation, say. The support of radix 2 to 16 is a feature limitation, and something we should cleanly fail on in any case, regardless of whether parameter validation is on.